presto-main/src/main/java/io/prestosql/operator/scalar/FormatNumberFunction.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import io.airlift.slice.Slice;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.StandardTypes;
+
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+
+import static io.airlift.slice.Slices.utf8Slice;
+import static java.lang.String.format;
+
+public final class FormatNumberFunction
+{
+    private FormatNumberFunction() {}
+
+    @Description("formats large number using a unit symbol")
+    @ScalarFunction
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice formatNumber(@SqlType(StandardTypes.BIGINT) long num)
+    {
+        return utf8Slice(formatCount(num));
+    }
+
+    @Description("formats large number using a unit symbol")
+    @ScalarFunction
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice formatNumber(@SqlType(StandardTypes.DOUBLE) double num)
+    {
+        return utf8Slice(formatCount((long) num));
+    }
+
+    private static String formatCount(long count)
+    {
+        double fractional = count;
+        String unit = "";
+        if (fractional >= 1000 || fractional <= -1000) {
+            fractional /= 1000;
+            unit = "K";
+        }
+        if (fractional >= 1000 || fractional <= -1000) {
+            fractional /= 1000;
+            unit = "M";
+        }
+        if (fractional >= 1000 || fractional <= -1000) {
+            fractional /= 1000;
+            unit = "B";
+        }
+        if (fractional >= 1000 || fractional <= -1000) {
+            fractional /= 1000;
+            unit = "T";
+        }
+        if (fractional >= 1000 || fractional <= -1000) {
+            fractional /= 1000;
+            unit = "Q";
+        }
+
+        return format("%s%s", getFormat(fractional).format(fractional), unit);
+    }
+
+    private static DecimalFormat getFormat(double value)
+    {
+        DecimalFormat format;
+        if (value < 10) {
+            // show up to two decimals to get 3 significant digits
+            format = new DecimalFormat("#.##");
```

Comment:
This can be expensive to do for every row. Unfortunately, the instances are not thread-safe, so the can't be reused across threads. We should create them and store them in thread-local variables.

Reply:
Instead of thread local, make the function non-static and store the state in the class. The engine will create instances as needed.

Reply:
@electrum Do you mean creating an inner class?

Reply:
No, take a look at `ArrayShuffleFunction`. I couldn’t find a non-parameterized example, but for this you can use a no-arg constructor (the default one is fine).

We can probably implement this by having the three `DecimalFormat` variants created as instance variables and picking the right one here. It’s effectively caching them but simpler since they’re fixed.

Reply:
Got it, thanks for providing the information.